File: disk_speed.rb
# disk_speed.rb — Ruby версия

require 'sys/filesystem'
require 'sys/proctable'
require 'json'
require 'time'
require 'io/console'

class DiskSpeedMonitor
  def initialize(interval = 2)
    @interval = interval
    @stats = {}
    @prev_read = {}
    @prev_write = {}
  end

  def get_disk_io
    read = {}
    write = {}

    if RUBY_PLATFORM =~ /linux/
      # Linux: читаем /proc/diskstats
      File.readlines('/proc/diskstats').each do |line|
        parts = line.strip.split
        next if parts.length < 14
        # parts[2] - имя диска, parts[5] - read_sectors, parts[9] - write_sectors
        disk = parts[2]
        read_sectors = parts[5].to_i * 512
        write_sectors = parts[9].to_i * 512
        read[disk] = read_sectors
        write[disk] = write_sectors
      end
    elsif RUBY_PLATFORM =~ /darwin/
      # macOS: используем iostat
      output = `iostat -d 1 2 2>/dev/null`
      output.each_line do |line|
        parts = line.strip.split
        next if parts.length < 4 || parts[0] == 'disk'
        disk = parts[0]
        read[disk] = (parts[3].to_f * 1024).to_i
        write[disk] = (parts[4].to_f * 1024).to_i
      end
    else
      # Windows: заглушка
      drives = Sys::Filesystem.mounts.map(&:name)
      drives.each do |drive|
        read[drive] = 0
        write[drive] = 0
      end
    end

    [read, write]
  end

  def run
    puts "\e[36m💾 Disk Speed Monitor (Ruby)\e[0m"
    puts "Интервал: #{@interval} сек"
    puts "Нажмите Ctrl+C для остановки..."

    @prev_read, @prev_write = get_disk_io
    sleep @interval

    trap('INT') do
      puts "\n\n⏹️ Остановка..."
      print_stats
      exit 0
    end

    loop do
      curr_read, curr_write = get_disk_io
      speeds = calculate_speeds(@prev_read, @prev_write, curr_read, curr_write)
      print_speeds(speeds)
      @prev_read = curr_read
      @prev_write = curr_write
      sleep @interval
    end
  end

  def calculate_speeds(prev_r, prev_w, curr_r, curr_w)
    speeds = {}
    curr_r.each do |disk, read|
      if prev_r[disk] && curr_r[disk]
        read_mb = (read - prev_r[disk]) / (1024.0 * 1024.0)
        write_mb = (curr_w[disk] - prev_w[disk]) / (1024.0 * 1024.0)
        read_speed = read_mb / @interval
        write_speed = write_mb / @interval

        speeds[disk] = { read: read_speed, write: write_speed }

        @stats[disk] ||= { read: [], write: [], max_read: 0, max_write: 0 }
        @stats[disk][:read] << read_speed
        @stats[disk][:write] << write_speed
        @stats[disk][:max_read] = read_speed if read_speed > @stats[disk][:max_read]
        @stats[disk][:max_write] = write_speed if write_speed > @stats[disk][:max_write]
      end
    end
    speeds
  end

  def print_speeds(speeds)
    puts "\n" + "─" * 60
    puts "\e[36mДиск   Чтение (МБ/с)   Запись (МБ/с)   Загрузка\e[0m"
    puts "─" * 60

    speeds.each do |disk, speed|
      read = speed[:read] || 0
      write = speed[:write] || 0
      load = [(read + write) / 1000 * 100, 100].min

      read_color = read > 300 ? "\e[32m" : read > 100 ? "\e[33m" : "\e[31m"
      write_color = write > 300 ? "\e[32m" : write > 100 ? "\e[33m" : "\e[31m"

      bar_len = 20
      filled = (load / 100 * bar_len).to_i
      bar = "█" * filled + "░" * (bar_len - filled)

      puts "#{disk.to_s.ljust(6)} #{read_color}#{read.round(1).to_s.rjust(8)}\e[0m     #{write_color}#{write.round(1).to_s.rjust(8)}\e[0m     #{bar} #{load.round.to_s.rjust(5)}%"
    end
  end

  def print_stats
    puts "\n\e[36m📊 Статистика:\e[0m"
    @stats.each do |disk, stat|
      if stat[:read].any?
        avg_read = stat[:read].sum / stat[:read].size
        avg_write = stat[:write].sum / stat[:write].size
        puts "  #{disk}:"
        puts "    Средняя скорость чтения: #{avg_read.round(1)} МБ/с"
        puts "    Средняя скорость записи: #{avg_write.round(1)} МБ/с"
        puts "    Пиковая скорость чтения: #{stat[:max_read].round(1)} МБ/с"
        puts "    Пиковая скорость записи: #{stat[:max_write].round(1)} МБ/с"
      end
    end
  end
end

interval = (ARGV[0] || 2).to_i
monitor = DiskSpeedMonitor.new(interval)
monitor.run

File: disk_speed.php
<?php
// disk_speed.php — PHP версия

class DiskSpeedMonitor {
    private $interval;
    private $stats = [];
    private $prevRead = [];
    private $prevWrite = [];

    public function __construct($interval = 2) {
        $this->interval = $interval;
    }

    private function getDiskIO() {
        $read = [];
        $write = [];

        if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN') {
            // Windows: используем wmic
            $output = shell_exec('wmic diskdrive get DeviceID,Name,Size 2>nul');
            // Упрощённо: только системный диск
            $read['C:'] = 0;
            $write['C:'] = 0;
        } else {
            // Linux/macOS: используем iostat
            $output = shell_exec('iostat -d 1 2 2>/dev/null');
            $lines = explode("\n", $output);
            foreach ($lines as $line) {
                $parts = preg_split('/\s+/', trim($line));
                if (count($parts) >= 7 && is_numeric($parts[5])) {
                    $disk = $parts[0];
                    $read[$disk] = (int)($parts[5] * 1024);
                    $write[$disk] = (int)($parts[6] * 1024);
                }
            }
        }
        return [$read, $write];
    }

    private function calculateSpeeds($prevR, $prevW, $currR, $currW) {
        $speeds = [];
        foreach ($currR as $disk => $read) {
            if (isset($prevR[$disk])) {
                $readMB = ($read - $prevR[$disk]) / (1024 * 1024);
                $writeMB = ($currW[$disk] - $prevW[$disk]) / (1024 * 1024);
                $readSpeed = $readMB / $this->interval;
                $writeSpeed = $writeMB / $this->interval;

                $speeds[$disk] = ['read' => $readSpeed, 'write' => $writeSpeed];

                if (!isset($this->stats[$disk])) {
                    $this->stats[$disk] = ['read' => [], 'write' => [], 'max_read' => 0, 'max_write' => 0];
                }
                $this->stats[$disk]['read'][] = $readSpeed;
                $this->stats[$disk]['write'][] = $writeSpeed;
                if ($readSpeed > $this->stats[$disk]['max_read']) {
                    $this->stats[$disk]['max_read'] = $readSpeed;
                }
                if ($writeSpeed > $this->stats[$disk]['max_write']) {
                    $this->stats[$disk]['max_write'] = $writeSpeed;
                }
            }
        }
        return $speeds;
    }

    private function printSpeeds($speeds) {
        echo "\n" . str_repeat("─", 60) . "\n";
        echo "\033[36mДиск   Чтение (МБ/с)   Запись (МБ/с)   Загрузка\033[0m\n";
        echo str_repeat("─", 60) . "\n";

        foreach ($speeds as $disk => $speed) {
            $read = $speed['read'] ?? 0;
            $write = $speed['write'] ?? 0;
            $load = min(($read + $write) / 1000 * 100, 100);

            $readColor = $read > 300 ? "\033[32m" : ($read > 100 ? "\033[33m" : "\033[31m");
            $writeColor = $write > 300 ? "\033[32m" : ($write > 100 ? "\033[33m" : "\033[31m");

            $barLen = 20;
            $filled = (int)($load / 100 * $barLen);
            $bar = str_repeat("█", $filled) . str_repeat("░", $barLen - $filled);

            printf("%-6s %s%8.1f\033[0m     %s%8.1f\033[0m     %s %5.0f%%\n",
                $disk, $readColor, $read, $writeColor, $write, $bar, $load);
        }
    }

    private function printStats() {
        echo "\n\033[36m📊 Статистика:\033[0m\n";
        foreach ($this->stats as $disk => $stat) {
            if (!empty($stat['read'])) {
                $avgRead = array_sum($stat['read']) / count($stat['read']);
                $avgWrite = array_sum($stat['write']) / count($stat['write']);
                echo "  $disk:\n";
                echo "    Средняя скорость чтения: " . round($avgRead, 1) . " МБ/с\n";
                echo "    Средняя скорость записи: " . round($avgWrite, 1) . " МБ/с\n";
                echo "    Пиковая скорость чтения: " . round($stat['max_read'], 1) . " МБ/с\n";
                echo "    Пиковая скорость записи: " . round($stat['max_write'], 1) . " МБ/с\n";
            }
        }
    }

    public function run() {
        echo "\033[36m💾 Disk Speed Monitor (PHP)\033[0m\n";
        echo "Интервал: {$this->interval} сек\n";
        echo "Нажмите Ctrl+C для остановки...\n";

        pcntl_signal(SIGINT, function() {
            echo "\n\n⏹️ Остановка...\n";
            $this->printStats();
            exit(0);
        });

        list($this->prevRead, $this->prevWrite) = $this->getDiskIO();
        sleep($this->interval);

        while (true) {
            pcntl_signal_dispatch();
            list($currRead, $currWrite) = $this->getDiskIO();
            $speeds = $this->calculateSpeeds($this->prevRead, $this->prevWrite, $currRead, $currWrite);
            $this->printSpeeds($speeds);
            $this->prevRead = $currRead;
            $this->prevWrite = $currWrite;
            sleep($this->interval);
        }
    }
}

$interval = isset($argv[1]) ? (int)$argv[1] : 2;
$monitor = new DiskSpeedMonitor($interval);
$monitor->run();
?>
